### DISCUSSION TITLE
Integrating Bentoml With Jenkins

### DISCUSSION THREAD

**Discussion by dummybot88** [bentoml/BentoML#1280] [Q&A]
Hi,

I'm trying to integrate bentoml with Jenkins and deployed the same on an AWS EC2 instance. If I run the bentoml lambda or sagemaker commands to deploy the bento service from Jenkins Pipeline job, I'm facing the below exception

`bentoml lambda deploy second-15 -b IrisClassifier:20201127123143_1828D1
Error: lambda deploy failed: INTERNAL:BentoML currently not support UNSET repository`

`bentoml lambda deploy second-12 -b IrisClassifier:20201127094323_D3BC84
Deploying "IrisClassifier:20201127094323_D3BC84" to AWS Lambda [2020-11-27 12:00:28,634] ERROR - URPC ERROR ApplyDeployment: None has type NoneType, but expected one of: bytes, unicode
Error: lambda deploy failed: INTERNAL:None has type NoneType, but expected one of: bytes, unicode`

Could you please throw some light on how to fix the issue. Below are the versions of tools involved:
Python: 3.7.9
bentoml: 0.9.2
aws cli: 2.1.4
aws sam cli: 0.33.1
docker: 19.03.13-ce

**Comment by yubozhao** [comment]
@dummybot88 Can you share with me how did you setup the Jenkin job?  Also can you run `bentoml get IrisClassifier:20201127123143_1828D1` command and share the output with me? This will help me understand how did you saved the Bento bundle better.

**Reply by yubozhao** [reply]
@dummybot88 Let me know if the issue still happens for you. Happy to help

**Reply by dummybot88** [reply]
Hey @yubozhao , thank you for the reply. The issue is resolved. I had to configure AWS default region and create an IAM role with required permissions to deploy the bento service to AWS lambda.

I would like to pose another question here. Does bentoml support ML feedback loops? I see in the documentation about feedback logs, but I don't think that does the task of feedback loop. Could you please throw some light here.

**Reply by yubozhao** [reply]
@dummybot88 Apology on the delayed message.

BentoML provides a way for you to collect feedback on the predicted results.  You can send feedback to the `/feedback` endpoint in json format with the `request_id` you get in the prediction result return.  BentoML will then stores that feedback in the feedback log file. Users will need to handle how to incorporate this into their workflow.

This feature still needs a lot of improvements. I'd love to hear your current workflow setup for the feedback loop.

**Reply by dummybot88** [reply]
Thank you for the elaborated response. I would like to extend my question further. Do we have a provision to override the default log directory and point it to an S3 bucket. I want all the log files to sit in s3 bucket instead of the local machine. 

I tried to update the `base_log_dir` value in the config file to point to s3 and observed that bentoml has created a directory structure with the S3 URL within the local machine and placed the log files inside that directory.

### Default Value:
`base_log_dir = {bentoml_home}/logs`

### Overridden Value:
`base_log_dir = s3://testingbentoml/logs/`

Here, bentoml is creating a directory `s3:` with `testingbentoml `and `logs` as sub-directories instead of writing the log files to the s3 bucket. 

Can you please let me know if there is a feature in place to write the log files to a remote location like s3?

**Reply by yubozhao** [reply]
@dummybot88 BentoML use python's stand logging library.  So you can customizes it base on your needs.  

For log to s3, you would need to install new handler such as: https://pypi.org/project/aws-logging-handlers.
We have a PR ongoing that is providing an easier access to the full log customization.

Right now, a quick workaround for you without waiting for the next release. You can try adding the following line in your BentoService class definition file, after importing BentoML:

```
import bentoml
import logging
from aws_logging_handlers.S3 import S3Handler
s3_handler = S3Handler("test_log", bucket, workers=3)
logging.getLogger('bentoml').addHandler(s3_handler)
```

**Reply by dummybot88** [reply]
Thank you @yubozhao, I'll give it a try

**Reply by dummybot88** [reply]
Hello ,

I'm facing an issue while trying to deploy bento service saved in S3 as `tar.gz` file to AWS Lambda.

`[2021-01-07 11:59:38,249] WARNING - pip package requirement `bentoml==0.10.1` not found in current python environment
[2021-01-07T17:29:38.277+05:30] /tmp/requirements/joblib/_multiprocessing_helpers.py:45: UserWarning: [Errno 38] Function not implemented.  joblib will operate in serial mode
[ERROR] Runtime.ImportModuleError: Unable to import module 'app': libgomp-f7e03b3e.so.1.0.0: cannot open shared object file: No such file or directory
[2021-01-07T17:29:38.373+05:30] It seems that scikit-learn has not been built correctly.
[2021-01-07T17:29:38.373+05:30] to build the package before using it: run `python setup.py install` or`

The above logs are observed in AWS cloud watch when trying to invoke the `predict` API.

Could anyone provide some inputs here?

**Reply by dummybot88** [reply]
Hello @yubozhao,

Regarding logging to S3. I have implemented the workaround suggested by you in the above comments in the quick start example. Please find below the bento service

```import pandas as pd

import bentoml
from bentoml import env, artifacts, api, BentoService
from bentoml.adapters import DataframeInput
from bentoml.frameworks.sklearn import SklearnModelArtifact
import logging
from aws_logging_handlers.S3 import S3Handler

@env(infer_pip_packages=True)
@artifacts([SklearnModelArtifact('model')])
class IrisClassifier(BentoService):
   
    bucket="testingbentoml"
    s3_handler = S3Handler("test_log", bucket, workers=3)
    logging.getLogger('bentoml').addHandler(s3_handler)

    @api(input=DataframeInput(), batch=True)
    def predict(self, df: pd.DataFrame):
        """
        An inference API named `predict` with Dataframe input adapter, which codifies
        how HTTP requests or CSV files are converted to a pandas Dataframe object as the
        inference API function input
        """
        logger.info("test info message")
        return self.artifacts.model.predict(df)`



And when I tried to run `python3 main.py`, I have encountered the below error, can you throw some light here:

[root@ip quick-start]# python3 main.py 
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aws_logging_handlers/S3/__init__.py", line 172, in _get_stream_object
    return StreamObject(self.s3, self.bucket.name, filename, self._stream_buffer_queue, self.encryption_options)
  File "/usr/local/lib/python3.7/site-packages/aws_logging_handlers/S3/__init__.py", line 34, in __init__
    self.uploader = self.object.initiate_multipart_upload(**encryption)
TypeError: initiate_multipart_upload() argument after ** must be a mapping, not NoneType

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "main.py", line 4, in <module>
    from iris_classifier import IrisClassifier
  File "/home/ec2-user/bentoml/guides/quick-start/iris_classifier.py", line 12, in <module>
    class IrisClassifier(BentoService):
  File "/home/ec2-user/bentoml/guides/quick-start/iris_classifier.py", line 17, in IrisClassifier
    s3_handler = S3Handler("test_log", bucket, workers=3)
  File "/usr/local/lib/python3.7/site-packages/aws_logging_handlers/S3/__init__.py", line 331, in __init__
    compress=compress, **boto_session_kwargs)
  File "/usr/local/lib/python3.7/site-packages/aws_logging_handlers/S3/__init__.py", line 109, in __init__
    self._current_object = self._get_stream_object(self.current_file_name)
  File "/usr/local/lib/python3.7/site-packages/aws_logging_handlers/S3/__init__.py", line 175, in _get_stream_object
    raise RuntimeError('Failed to open new S3 stream object')
RuntimeError: Failed to open new S3 stream object


Thanks,
Sravan

**Reply by yubozhao** [reply]
I suggest you have the handler definition and update logging handler statement outside of the BentoService class definition and try again.

From the aws_logging_handlers library's documentation, it failed to initialize their `StreamObject` https://github.com/omrikiei/aws_logging_handlers/blob/448c23d17f540b0b0c39461512a36fcc50bcdb46/aws_logging_handlers/S3/__init__.py#L172

You probably can find better answers in in there.
